File: go/matmul.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func llenarMatrices(A []float64, B []float64, n int) {
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			A[i*n+j] = float64((i+2*j)%100) / 20.0
			B[i*n+j] = float64((3*i+j)%100) / 30.0
		}
	}
}

func limpiar(C []float64) {
	for i := range C {
		C[i] = 0.0
	}
}

func multiplicacionClasica(A []float64, B []float64, C []float64, n int) float64 {
	limpiar(C)

	inicio := time.Now()
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			suma := 0.0
			for k := 0; k < n; k++ {
				suma += A[i*n+k] * B[k*n+j]
			}
			C[i*n+j] = suma
		}
	}

	return time.Since(inicio).Seconds() * 1000.0
}

func minimo(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

func multiplicacionBloques(A []float64, B []float64, C []float64, n int, bloque int) float64 {
	limpiar(C)

	inicio := time.Now()
	for ii := 0; ii < n; ii += bloque {
		for jj := 0; jj < n; jj += bloque {
			for kk := 0; kk < n; kk += bloque {
				finI := minimo(ii+bloque, n)
				finJ := minimo(jj+bloque, n)
				finK := minimo(kk+bloque, n)

				for i := ii; i < finI; i++ {
					for j := jj; j < finJ; j++ {
						suma := C[i*n+j]
						for k := kk; k < finK; k++ {
							suma += A[i*n+k] * B[k*n+j]
						}
						C[i*n+j] = suma
					}
				}
			}
		}
	}

	return time.Since(inicio).Seconds() * 1000.0
}

func checksum(C []float64) float64 {
	suma := 0.0
	for _, valor := range C {
		suma += valor
	}
	return suma
}

func main() {
	tamanos := []int{64, 128, 256, 384}
	bloque := 32

	if len(os.Args) > 1 {
		tamanos = nil
		for i := 1; i < len(os.Args); i++ {
			arg := os.Args[i]
			if (arg == "--block" || arg == "-b") && i+1 < len(os.Args) {
				if b, err := strconv.Atoi(os.Args[i+1]); err == nil && b > 0 {
					bloque = b
				}
				i++
				continue
			}
			if strings.HasPrefix(arg, "--block=") {
				val := strings.TrimPrefix(arg, "--block=")
				if b, err := strconv.Atoi(val); err == nil && b > 0 {
					bloque = b
				}
				continue
			}
			n, err := strconv.Atoi(arg)
			if err == nil && n > 0 {
				tamanos = append(tamanos, n)
			}
		}
	}

	fmt.Println("Comparacion de multiplicacion de matrices")
	fmt.Printf("Bloque usado en la version por bloques: %d\n", bloque)
	fmt.Printf("%-10s%-20s%-20s%-14s%s\n", "N", "clasica (ms)", "bloques (ms)", "razon", "checksum")

	for _, n := range tamanos {
		A := make([]float64, n*n)
		B := make([]float64, n*n)
		C := make([]float64, n*n)

		llenarMatrices(A, B, n)

		tiempoClasico := multiplicacionClasica(A, B, C, n)
		control := checksum(C)

		tiempoBloques := multiplicacionBloques(A, B, C, n, bloque)
		control += checksum(C)

		fmt.Printf("%-10d%-20.3f%-20.3f%-14.2f%.2f\n", n, tiempoClasico, tiempoBloques, tiempoClasico/tiempoBloques, control)
	}
}

File: Tarea Cache/matmul.cpp
#include <algorithm>
#include <chrono>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <vector>

using namespace std;
using reloj = chrono::high_resolution_clock;

static void llenar_matrices(vector<double>& A, vector<double>& B, int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            A[(size_t)i * n + j] = (double)((i + 2 * j) % 100) / 20.0;
            B[(size_t)i * n + j] = (double)((3 * i + j) % 100) / 30.0;
        }
    }
}

static void limpiar(vector<double>& C) {
    fill(C.begin(), C.end(), 0.0);
}

static double multiplicacion_clasica(const vector<double>& A, const vector<double>& B, vector<double>& C, int n) {
    limpiar(C);

    auto inicio = reloj::now();
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            double suma = 0.0;
            for (int k = 0; k < n; k++) {
                suma += A[(size_t)i * n + k] * B[(size_t)k * n + j];
            }
            C[(size_t)i * n + j] = suma;
        }
    }
    auto fin = reloj::now();

    return chrono::duration<double, milli>(fin - inicio).count();
}

static double multiplicacion_bloques(const vector<double>& A, const vector<double>& B, vector<double>& C, int n, int bloque) {
    limpiar(C);

    auto inicio = reloj::now();
    for (int ii = 0; ii < n; ii += bloque) {
        for (int jj = 0; jj < n; jj += bloque) {
            for (int kk = 0; kk < n; kk += bloque) {
                int fin_i = min(ii + bloque, n);
                int fin_j = min(jj + bloque, n);
                int fin_k = min(kk + bloque, n);

                for (int i = ii; i < fin_i; i++) {
                    for (int j = jj; j < fin_j; j++) {
                        double suma = C[(size_t)i * n + j];
                        for (int k = kk; k < fin_k; k++) {
                            suma += A[(size_t)i * n + k] * B[(size_t)k * n + j];
                        }
                        C[(size_t)i * n + j] = suma;
                    }
                }
            }
        }
    }
    auto fin = reloj::now();

    return chrono::duration<double, milli>(fin - inicio).count();
}

static double checksum(const vector<double>& C) {
    double suma = 0.0;
    for (double valor : C) {
        suma += valor;
    }
    return suma;
}

int main(int argc, char* argv[]) {
    vector<int> tamanos;
    int bloque = 32;

    if (argc > 1) {
        for (int i = 1; i < argc; i++) {
            tamanos.push_back(atoi(argv[i]));
        }
    } else {
        tamanos = {64, 128, 256, 384};
    }

    cout << "Comparacion de multiplicacion de matrices" << '\n';
    cout << "Bloque usado en la version por bloques: " << bloque << '\n';
    cout << left << setw(10) << "N"
         << setw(20) << "clasica (ms)"
         << setw(20) << "bloques (ms)"
         << setw(14) << "razon"
         << "checksum" << '\n';

    for (int n : tamanos) {
        vector<double> A((size_t)n * n);
        vector<double> B((size_t)n * n);
        vector<double> C((size_t)n * n);

        llenar_matrices(A, B, n);

        double tiempo_clasico = multiplicacion_clasica(A, B, C, n);
        double control = checksum(C);

        double tiempo_bloques = multiplicacion_bloques(A, B, C, n, bloque);
        control += checksum(C);

        cout << left << setw(10) << n
             << setw(20) << fixed << setprecision(3) << tiempo_clasico
             << setw(20) << fixed << setprecision(3) << tiempo_bloques
             << setw(14) << fixed << setprecision(2) << (tiempo_clasico / tiempo_bloques)
             << fixed << setprecision(2) << control << '\n';
    }

    return 0;
}
